Add `tuist inspect implicit-dependencies` to detect implicit dependencies

rofle100lvl opened the pull request:
Part of https://github.com/tuist/tuist/issues/6492

### Short description 📝

> Describe here the purpose of your PR.

The purpose of this PR is to find way to get all the imports of the module to check it later on implicitness. How @kwridan noticed in [comment](https://github.com/tuist/tuist/issues/6492#issuecomment-2213413457), the best way to do it is static analysis and pattern matching. Also I added ImportFinderCommand in purpose of testing ImportFinderService. I tested it on Tuist project and it works, but still a lot of things unclear, for example, targets with the same names and other corner cases.

### How to test the changes locally 🧐

Launch import_finder command with --path and --target-name options. For example, 
 tuist lint implicit-imports --path tuist/fixtures/ios_app_with_frameworks

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/get-started) for reference).

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

rofle100lvl commented:
Do not forget about @testable import

pepicrft commented on Sources/TuistKit/Commands/ImportFinderCommand.swift:
```diff
@@ -0,0 +1,39 @@
+import Path
+import TuistSupport
+
+public struct ImportFinderCommand: AsyncParsableCommand {
```
I'd suggest removing this one as a command that users directly interface with. I think that command should be `tuist lint implicit-imports` instead, which can be implemented in a follow-up PR.

rofle100lvl replied:
I agree with you, but now I need a way to test it. So I remove this command, when we will be ready to merge

pepicrft replied:
Sounds good! Thanks for clarifying @rofle100lvl

pepicrft commented on Sources/TuistKit/Services/ImportFinderService.swift:
```diff
@@ -0,0 +1,105 @@
+    }
+}
+
+final class ImportFinderService {
```
I'd name this one `GraphImplicitImportLintService` and make it more generic to work with the entire graph. What do you think? I think it's more representative of what the service does.

rofle100lvl replied:
Done

pepicrft commented on Sources/TuistKit/Services/ImportFinderService.swift:
```diff
@@ -0,0 +1,105 @@
+        path: AbsolutePath,
+        targetName: String
+    ) async throws {
+        let (graph, _, _) = try await manifestGraphLoader.load(path: path)
+        let target = graph
+            .projects
+            .map(\.value.targets)
+            .flatMap { $0 }
+            .map(\.value)
+            .filter { $0.name == targetName }
+            .first
+        guard let target else { throw ImportFinderServiceError.targetNotFound(target: targetName) }
+        let imports = Set(await handleTarget(target: target))
+        print("Used imports: \n\(imports.joined(separator: ", "))")
```
I'd lint the entire graph, and then output something along the lines of:
```
The following implicit imports have been detected through static code analysis:
  - Target A implicitly imports A, B, and C.
  - Target B implicitly imports X, Y, Z
```

kwridan replied:
To avoid listing system frameworks or libraries we can further filter the results with all known targets from the graph. 

e.g. `UIKit` isn't part of the graph as one of the known targets, and as such would be excluded from the final output.

pepicrft replied:
Exactly. You don't want to see the output being noisy with the system frameworks.

rofle100lvl replied:
Now if you launch 
`tuist import_finder --path /Users/rofle100lvl/Desktop/tuist/fixtures/ios_app_with_framework_and_disabled_resources`
You will see

> The following implicit imports have been detected through static code analysis:
> Target App imports: StaticFramework, Framework1

Which is exactly what we want to see

pepicrft commented on Sources/TuistKit/Services/ImportFinderService.swift:
```diff
@@ -0,0 +1,105 @@
+
+    func handleTarget(target: XcodeGraph.Target) async -> [String] {
+        return await withTaskGroup(of: [String].self) { [weak self] group in
+            var imports = [String]()
```
Is this free of race conditions?

rofle100lvl replied:
Yes, it is

pepicrft commented on Sources/TuistKit/Services/ImportFinderService.swift:
```diff
@@ -0,0 +1,105 @@
+                language: language
+            )
+        } catch {
+            return []
```
This might end up hiding actual bugs. I'd bubble that up.

rofle100lvl replied:
Done

rofle100lvl commented on Sources/TuistKit/Utils/ImportSourceCodeScanner.swift:
```diff
@@ -0,0 +1,56 @@
+    case objc
+}
+
+final class ImportSourceCodeScanner {
```
Thank you for the suggestion. I have already implemented some tests in the ImportFinderTests file, and I'm considering whether to add more tests of the same type or to introduce other types of tests. If other, what kind of test would be meaningful here?

pepicrft replied:
I think those tests are fine!

kwridan commented on Tests/TuistKitTests/ImportFinder/ImportFinderTests.swift:
```diff
@@ -0,0 +1,75 @@
+
+@testable import TuistKit
+
+final class ImportFinderTests: TuistUnitTestCase {
```
nice - those tests are quite handy 👌

A few additional cases to consider are

* Swift multiple single line imports

```swift
import ModuleA; import ModuleB
```

* Swift import explicit clang modules

```
import ModuleA.Submodule
```

* Swift type imports

```swift
import struct ModuleA.SomeStruct
import enum ModuleA.SomeEnum
import class ModuleA.SomeClass
```

* Clang multiple single line imports

```objc
@import ModuleA; @import ModuleB;
```

* clang import explicit modules

```objc
@import ModuleA.Submodule;
```

rofle100lvl replied:
I added test to cover them all. If i forget something, tell me please

kwridan reviewed (commented):
Thanks for contributing this @rofle100lvl

rofle100lvl commented on Sources/TuistKit/Services/ImportFinderService.swift:
```diff
@@ -0,0 +1,105 @@
+        switch source.path.url.pathExtension {
+        case "swift":
+            language = .swift
+        case "objc":
```
Added

rofle100lvl commented:
I encounter an issue. If we want to use just regular expressions, we have that ones corner cases.
1. // import A
2. var a = "import B"
3. /*
import C
*/
And all of that cases regular expression will parse as true. Yes, we can handle 1 and 2 cases if we change expression(make it much more harder), but I have no idea how to fix third one. 
Maybe here we should look in SwiftSyntax side, yes it's slower but we can be sure that this import is import statement and not a comment or a string. 
Maybe for demo version regular expressions are enough, but we need to consider that enhancement

rofle100lvl commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,36 @@
+
+@testable import TuistKit
+
+final class GraphImplicitImportLintServiceTests: TuistUnitTestCase {
```
Hey, @pepicrft. I cannot understand what is the best way to do this test. It doesn't work for me. As far as I researched all code for oppening graph 
`let manifestLoader = ManifestLoaderFactory()
            .createManifestLoader()
        let manifestGraphLoader = ManifestGraphLoader(
            manifestLoader: manifestLoader,
            workspaceMapper: SequentialWorkspaceMapper(mappers: []),
            graphMapper: SequentialGraphMapper([])
        )`
 was never tested together. What is the best way to open project?

pepicrft replied:
@rofle100lvl since the graph takes a `Graph` as an input and throws or returns based on the graph, I'd suggest that your tests instantiate an in-memory graph and then use it as an object under test:

```swift
// Instantiate the graph
let graph = Graph(....)

// Assert
XCT....(try await GraphImplicitImportLintService(graph: graph).lint())
```

Because the graph will reference real files, I recommend creating them in a temporary directory. Every test case gets a temporary directory assigned that you can use for that purpose. You can get it by calling: `let temporaryPath = try temporaryPath()`. 

You can use [this test](https://github.com/tuist/tuist/blob/f2501a9cb5bb9a416f19af97d605da447e705b8b/Tests/TuistGeneratorTests/Linter/GraphLinterTests.swift#L32) as a reference.

pepicrft replied:
The approach that you are currently taking integrate-tests the loading logic, which is already tested elsewhere, so I'd not do that as part of this test.

rofle100lvl replied:
@pepicrft thank you for your advice. I made tests to check the functionality. I think I'm ready to merge it and start making tuist lint command with matching target and names

rofle100lvl commented:
After [tests](https://github.com/tuist/tuist/pull/6603) on tuist repository. I found two issues now. 
1) How to check only tuist related projects and not dependencies? 
You can see in the log after all implicit dependencies were removed that dependencies of Tuist have implicit dependencies as well. 
2) You can see 
> Target ProjectDescription implicitly imports ProjectDescription
It appears in code because of this [line](https://github.com/tuist/tuist/blob/4c16b9a712539bb2163c42facd141b9944063961/Sources/ProjectDescription/Config.swift#L23)
But for other cases it works brilliant

kwridan commented on Sources/TuistKit/Commands/ImplicitImportsLintCommand.swift:
```diff
@@ -0,0 +1,52 @@
+        let (graph, _, _, _) = try await manifestGraphLoader
+            .load(path: projectPath)
+        for (target, implicitDependencies) in try await GraphImplicitImportLintService(graph: graph).lint() {
+            print("Target \(target.name) implicitly imports \(implicitDependencies.joined(separator: ", ")).")
```
A warning can be emitted instead of a print - this can help give it the correct formatting in the command line.

```suggestion
            logger.warning("Target \(target.name) implicitly imports \(implicitDependencies.joined(separator: ", ")).")
```

kwridan reviewed (commented):
thanks @rofle100lvl

kwridan reviewed (approved):
Awesome thank you @rofle100lvl 🙏

Will have a think about the comments false positive case, indeed it's a shortcoming of this approach. It was a trade off of speed & dependency overhead vs correctness of SwiftSyntax (and the need for something different for Objective-C) 😞 

Some ideas (can be experimented with in separate follow ups):

- Strip out lines with comments ahead of scanning for imports
- Have separate pattern for finding comments and filter out any identified imports that have ranges within the comment ranges
- implementing a custom scanner/parser rather than regex that would skip over comments

rofle100lvl commented:
@pepicrft Waiting for your review. After that I can make documentation about that. And make blog post as well, if it's okay for you. I would love to do it

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+            filesToScan.append(contentsOf: headers.public)
+            filesToScan.append(contentsOf: headers.project)
+        }
+        return try await withThrowingTaskGroup(of: [String].self) { [weak self] group in
```
Curious why we have not opted to use [concurrentMap](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Extensions/Array%2BExecutionContext.swift#L22) here?

rofle100lvl replied:
You are right. It's more readable

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+    var errorDescription: String? {
+        switch self {
+        case .readFile:
+            "Error while reading file"
```
Should we include the name of the file that we failed reading?

rofle100lvl replied:
Removed that error. Now Filehandler will notify about it

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+            for target in allTargets {
+                let targetImports = Set(try await handleTarget(target: target))
+                let targetTuistDeclaredDependencies = target.dependencies.compactMap {
+                    var dependencyName: String?
+                    switch $0 {
+                    case let .target(name: targetName, _):
+                        dependencyName = project.targets[targetName]?.productName
+                    case let .project(target: targetName, path: projectPath, _):
+                        dependencyName = graph.projects[projectPath]?.targets[targetName]?.productName
+                    default:
+                        break
+                    }
+                    return dependencyName
```
```suggestion
                    switch $0 {
                    case let .target(name: targetName, _):
                        return project.targets[targetName]?.productName
                    case let .project(target: targetName, path: projectPath, _):
                        return graph.projects[projectPath]?.targets[targetName]?.productName
                    default:
                        return nil
                    }
```

We don't really use the local variable for anything, so we can just immediately return

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+                    }
+                    return dependencyName
+                }
+                var implicitImports: Set<String> = []
+                for targetImport in targetImports {
+                    if allTargetNames.contains(targetImport), !targetTuistDeclaredDependencies.contains(targetImport) {
+                        implicitImports.insert(targetImport)
+                    }
+                }
```
```suggestion
                let implicitImports = targetImports.intersection(allTargetNames).subtracting(targetTuistDeclaredDependencies)
```

This _should_ be equivalent and in my mind more readable, but could be just me 😆

rofle100lvl replied:
Yep, you are right. For me as well

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+                        implicitImports.insert(targetImport)
+                    }
+                }
+                if implicitImports.count > 0 {
```
```suggestion
                if !implicitImports.isEmpty {
```

nit

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+        return implicitTargetImports
+    }
+
+    func handleTarget(target: XcodeGraph.Target) async throws -> Set<String> {
```
```suggestion
    func imports(for target: XcodeGraph.Target) async throws -> Set<String> {
```

What do you think of making it more clear what the method does?

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+
+        let sourceCode = String(data: try FileHandler.shared.readFile(path), encoding: .utf8)
+        guard let sourceCode else { throw GraphImplicitImportLintServiceError.readFile }
+        return try ImportSourceCodeScanner().extractImports(
```
Should the `ImportSourceCodeScanner` be injected instead of creating on-demand here?

rofle100lvl replied:
Yep, now its injected

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+            return []
+        }
+
+        let sourceCode = String(data: try FileHandler.shared.readFile(path), encoding: .utf8)
```
You can use `readTextFile` here instead

rofle100lvl replied:
Done

fortmarek reviewed (commented):
Thanks a lot for the PR 👏 

Apart from the comments, I'd also add an extra acceptance test to test this flow E2E.

rofle100lvl commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+}
+
+final class GraphImplicitImportLintService {
```
GraphLinting protocol doesn't support concurrency. What do you think about that?

fortmarek replied:
Then I'd at least copy the function signature but with `async` without directly conforming to the protocol.

rofle100lvl replied:
Done

rofle100lvl commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
Well, if we want to divide Generator time lintning and Separated lintning it is not a good idea to store them in one module

rofle100lvl replied:
However I agree with u. That it should be done. Maybe we can make it in follow up pr? What do you think? Honestly, it's starting to be hard to keep all the context in this pr

rofle100lvl commented on Sources/TuistKit/Commands/ImplicitImportsLintCommand.swift:
```diff
@@ -0,0 +1,52 @@
+            .load(path: projectPath)
+        for (target, implicitDependencies) in try await GraphImplicitImportLintService(graph: graph).lint() {
+            logger.warning("Target \(target.name) implicitly imports \(implicitDependencies.joined(separator: ", ")).")
+        }
```
Separated logic in ImplicitImportsLintService

rofle100lvl commented on Sources/TuistKit/Commands/ImplicitImportsLintCommand.swift:
```diff
@@ -0,0 +1,52 @@
+            graphMapper: SequentialGraphMapper([])
+        )
+        let (graph, _, _, _) = try await manifestGraphLoader
+            .load(path: projectPath)
```
Done

rofle100lvl commented on Sources/TuistKit/Commands/LintCommand.swift:
```diff
@@ -0,0 +1,20 @@
+    public static var configuration: CommandConfiguration {
+        CommandConfiguration(
+            commandName: "lint",
+            subcommands: [ImplicitImportsLintCommand.self]
```
Made

rofle100lvl commented:
Added Acceptance test

fortmarek commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+@testable import TuistKit
+
+final class GraphImplicitImportLintServiceTests: TuistUnitTestCase {
+    func test_TargetLintWithImports() async throws {
```
I'd suggest making the test names a bit more clear in terms of the scenario – it's not clear to me how `test_TargetLintWithImports` is different from `test_TargetHandleWithImports`. I'd follow the same naming convention as we have in other tests (`test_target_lint_with_imports` or `test_targetLintWithImports`)

fortmarek commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+        let path = try temporaryPath()
+        let targetPath = path.appending(try RelativePath(validating: "FirstTarget"))
+
+        let targetFirstFile = targetPath.appending(component: "Sources").appending(component: "FirstFile.swift")
```
```suggestion
        let targetFirstFile = targetPath.appending(components: "Sources", "FirstFile.swift")
```

nit

fortmarek commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+
+    func test_TargetHandleWithImports() async throws {
+        let path = try temporaryPath()
+        let targetPath = path.appending(try RelativePath(validating: "FirstTarget"))
```
```suggestion
        let targetPath = path.appending(component: "FirstTarget")
```

fortmarek commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+        let targetFirstFile = targetPath.appending(component: "Sources").appending(component: "FirstFile.swift")
+        let targetSecondFile = targetPath.appending(component: "Sources").appending(component: "SecondFile.swift")
+
+        try FileHandler.shared.createFolder(targetPath)
```
```suggestion
        try fileHandler.createFolder(targetPath)
```

We're trying not to use the global `FileHandler.shared` instance as that will make it tricky to conform to the new Swift concurrency model.

I'd also try to use the new [FileSystem](https://github.com/tuist/FileSystem) instead of `FileHandler` wherever we can.

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    func imports(for target: XcodeGraph.Target) async throws -> Set<String> {
```
This should be a private method and not directly tested at all.

rofle100lvl replied:
But it is much more easier to make tests on this method rather than create tests for lint function. What do you think maybe move it to other Service? Because I strongly believe that this part should be tested

fortmarek replied:
I don't understand why we need to test the `imports` method separately. What matters is the input of `lint` and the output of lint issues. This should be an implementation detail.

rofle100lvl replied:
I want to separate it, because it's a big part of the feature. And it's easier to test it separately. Now I made TargetImportScanner, which contains all the logic with scanning targets

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,87 @@
+            return []
+        }
+
+        let sourceCode = try FileHandler.shared.readTextFile(path)
```
Let's not use the global `FileHandler.shared`. It should be accessed as a property of `GraphImplicitImportLintService`

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private func matchPattern(at path: AbsolutePath) async throws -> [String] {
+        var language: ProgrammingLanguage
```
```suggestion
        let language: ProgrammingLanguage
```

nit

fortmarek commented on Sources/TuistKit/Services/ImplicitImportsLintService.swift:
```diff
@@ -0,0 +1,31 @@
+import TuistCore
+import TuistLoader
+
+class ImplicitImportsLintService {
```
```suggestion
final class ImplicitImportsLintService {
```

nit

fortmarek commented on Sources/TuistKit/Services/ImplicitImportsLintService.swift:
```diff
@@ -0,0 +1,31 @@
+import TuistCore
+import TuistLoader
+
+class ImplicitImportsLintService {
```
Let's add some basic tests for this one

rofle100lvl replied:
What kind of tests do you want to add? LintImplisitCommand is tested and all services inside of it are tested. I can do similar to inner test or something else?

fortmarek replied:
We shouldn't be testing `LintImplicitCommand` at all in unit tests as all the business logic should be in the `LintImplicitService`.

rofle100lvl replied:
We tested all Services inside of LintImplicitService as well, we made Acceptance tests on LintImplicitCommand. I dont get what kind if tests can be done on LintImplicitService

fortmarek commented on Sources/TuistKit/Services/ImplicitImportsLintService.swift:
```diff
@@ -0,0 +1,31 @@
+        let lintingErrors = try await graphImplicitLintService.lint(graph: GraphTraverser(graph: graph), config: config)
+        lintingErrors.printWarningsIfNeeded()
+        if lintingErrors.isEmpty {
+            logger.log(level: .info, "Implicit dependencies were not found.")
```
```suggestion
            logger.log(level: .info, "We did not find any implicit dependencies in your project.")
```

I think this reads better, but very subjective, so feel free to keep as-is

fortmarek commented on Tests/TuistKitTests/ImportFinder/ImportSourceCodeScannerTests.swift:
```diff
@@ -0,0 +1,173 @@
+
+        @end
+        """
+        let imports = try ImportSourceCodeScanner().extractImports(from: code, language: .objc)
```
I'd create a `subject` property for `ImportSourceCodeScanner` as we do in other unit tests

fortmarek commented on Tests/TuistKitAcceptanceTests/ImplicitImportsLintCommandAcceptanceTests.swift:
Nit: I'd name this just `LintAcceptanceTests`.

fortmarek commented on Tests/TuistKitAcceptanceTests/ImplicitImportsLintCommandAcceptanceTests.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+import TuistAcceptanceTesting
+
+final class ImplicitImportsLintCommandAcceptanceTests: TuistAcceptanceTestCase {
```
```suggestion
final class LintImplicitImportsAcceptanceTests: TuistAcceptanceTestCase {
```

Now that I think about it, I'd also rename the `ImplicitImportsLintCommand` to `LintImplicitImportsCommand` as that follows how it's used in the CLI.

fortmarek commented on Tests/TuistKitAcceptanceTests/ImplicitImportsLintCommandAcceptanceTests.swift:
```diff
@@ -0,0 +1,10 @@
+import TuistAcceptanceTesting
+
+final class ImplicitImportsLintCommandAcceptanceTests: TuistAcceptanceTestCase {
+    func test_ios_app_with_frameworks() async throws {
```
Let's add one extra acceptance test that does find some implicit dependencies. We already have a fixture for that: https://github.com/tuist/tuist/tree/main/fixtures/ios_app_with_implicit_dependencies

rofle100lvl commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+            path: project,
+        ])
+        let result = try await GraphImplicitImportLintService(
+            importSourceCodeScanner: ImportSourceCodeScanner()
```
I came up with this approach on different tests examples. For instance, ProjectEditorMapperTests.swift:531. Should i change smth?

fortmarek replied:
we can keep it

pepicrft commented on Sources/TuistKit/Commands/LintImplicitImportsCommand.swift:
```diff
@@ -0,0 +1,27 @@
+    static var configuration: CommandConfiguration {
+        CommandConfiguration(
+            commandName: "implicit-imports",
+            abstract: "Find implicit imports in project"
```
```suggestion
            abstract: "Find implicit imports in Tuist projects."
```

pepicrft commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,50 @@
+            let allTargets = project.targets.values
+
+            for target in allTargets {
+                let targetImports = Set(try await targetScanner.imports(for: target))
+                let targetTuistDeclaredDependencies = target.dependencies.compactMap {
```
```suggestion
                let sourceDependencies = Set(try await targetScanner.imports(for: target))
                let explicitTargetDependencies = target.dependencies.compactMap {
```

pepicrft commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,50 @@
+                        return nil
+                    }
+                }
+                let implicitImports = targetImports.intersection(allTargetNames).subtracting(targetTuistDeclaredDependencies)
```
What happens if I import `X` where `X` is declared as a dependency with a pre-compiled XCFramework? Will that cause this to fail?

rofle100lvl replied:
I started to use allInternalTargets method here. Before that you could see imports which are marked as implicit, but they aren't defined in tuist manifest. Check it out [here](https://github.com/tuist/tuist/pull/6603)

pepicrft commented on Sources/TuistKit/Services/LintImplicitImportsService.swift:
```diff
@@ -0,0 +1,41 @@
+import TuistLoader
+import TuistSupport
+
+final class LintImplicitImportsService {
```
I find having two services `LintImplicitImportsService`, and `GraphImplicitImportLintService` a bit confusing. Are we going to lint implicit imports in an object other than a graph? If not, then I'd merge everything under `LintImplicitImportsService`.

rofle100lvl replied:
I think maybe we will use GraphImplicitImportLintService separately(but not sure). Also, now its good divided on layers, but yeah mb it's not required to have so many layers. lgtm idk

pepicrft commented on Sources/TuistKit/Services/LintImplicitImportsService.swift:
```diff
@@ -0,0 +1,41 @@
+        lintingErrors.printWarningsIfNeeded()
+        if lintingErrors.isEmpty {
+            logger.log(level: .info, "We did not find any implicit dependencies in your project.")
+        }
```
We should exit with a code other than 0 if there are implicit imports.

rofle100lvl replied:
Hm, but why should it be a zero? Command made the job it should and found all dependencies. I think 0 should be in case when command cannot be completed?

pepicrft commented on Tests/TuistKitAcceptanceTests/LintAcceptanceTests.swift:
```diff
@@ -0,0 +1,17 @@
+
+    func test_ios_app_with_implicit_dependencies() async throws {
+        try await setUpFixture(.iosAppWithImplicitDependencies)
+        try await run(LintImplicitImportsCommand.self)
```
I'd expect this command to error out if it finds implicit imports.

rofle100lvl replied:
I can understand why, but also somewhy i dont want to mark it as error. Because command works good and its not an error xd

fortmarek replied:
This should be an error. The same way linters usually return a non-zero exit code if they find critical issues in your code.

fortmarek replied:
Yes, you will need to throw an error 🙂

rofle100lvl replied:
I dont understand what you want me to do. In case of error it would look like that.
<img width="968" alt="image" src="https://github.com/user-attachments/assets/9907b4aa-c546-4933-8c5b-0cc31e5af5f4">
We can print Lint errors as error description as well, but all that text is not suit this case.

> If you think it's a legit issue, please file an issue including

rofle100lvl replied:
Now it looks like that
<img width="736" alt="Screenshot 2024-08-20 at 00 16 49" src="https://github.com/user-attachments/assets/459c3de0-0af5-4e50-b9ae-21b3fdc75624">

rofle100lvl replied:
@pepicrft @fortmarek  is that okay?

rofle100lvl replied:
@pepicrft What do you think?

rofle100lvl replied:
Added --strict flag now, it can be presented as an error and as an warning(Helpful when you want to present it in xcode for example)

pepicrft commented on Tests/TuistKitAcceptanceTests/LintAcceptanceTests.swift:
```diff
@@ -0,0 +1,17 @@
+    func test_ios_app_with_implicit_dependencies() async throws {
+        try await setUpFixture(.iosAppWithImplicitDependencies)
+        try await run(LintImplicitImportsCommand.self)
+        XCTAssertStandardOutput(pattern: "Target FrameworkA implicitly imports FrameworkB.")
```
How does this look if there's more than one target with implicit imports? Do we show a bullet list?

rofle100lvl replied:
For now, it works like that:

> Target TuistCacheIntegrationTests implicitly imports TuistSupportTesting, TuistCore, XcodeGraph, TuistHasher, TuistSupport.
> Target tuist implicitly imports TuistSupport, TuistLoader.

pepicrft commented on Tests/TuistKitTests/ImportFinder/GraphImplicitImportLintServiceTests.swift:
```diff
@@ -0,0 +1,72 @@
+
+final class GraphImplicitImportLintServiceTests: TuistUnitTestCase {
+    func test_targetLintWithImports() async throws {
+        let fileSystem = FileSystem()
+        // Given
```
```suggestion
        // Given
        let fileSystem = FileSystem()
```

pepicrft commented on Tests/TuistKitTests/ImportFinder/ImportSourceCodeScannerTests.swift:
```diff
@@ -0,0 +1,179 @@
+final class ImportSourceCodeScannerTests: TuistUnitTestCase {
+    var subject: ImportSourceCodeScanner!
+
+    override func setUp() {
+        subject = ImportSourceCodeScanner()
+    }
```
```suggestion
    override func setUp() {
        subject = ImportSourceCodeScanner()
        super.setUp()
    }
    override func tearDown() {
      subject = nil
      super.tearDown()
    }
```

pepicrft commented on docs/docs/guides/quick-start/add-dependencies.md:
```diff
@@ -120,4 +120,8 @@ struct ContentView_Previews: PreviewProvider {
\ No newline at end of file
+Run the app from Xcode, and you should see the image loaded from the URL.
+
+## Check your dependencies on implicitness
+
+Run the command `tuist lint implicit-imports` to lint the project for implicit imports. This command will scan all source files and display any implicit imports present in your codebase. To make an import explicit, simply add the imported module to the target in your Tuist project.
```
I'd not add this to a quick-start tutorial, which is focused on guiding the person throughout the most important features. What do you think if we add a new page under "Develop", "Validation", where the first page is "Lint".

pepicrft reviewed (changes requested):
This is amazing @rofle100lvl 👏🏼 👏🏼. Teams will love using this feature.

rofle100lvl commented:
I understand that it would be very nice to add --format  xcode option to launch this from xcode script. To show implicit imports as warnings directly in xcode. It's not very hard, I will add it during next PR

rofle100lvl commented:
Wow, guys. You need to check it out.
<img width="1172" alt="image" src="https://github.com/user-attachments/assets/a33497b4-b7d4-4b72-a228-486522603ecc">
I added xcode warnings, so now you can see warning directly in xcode. Also I added --strict flag to notify service does user want to see error or warnings(It's maden like that in other lintners(Periphery, SwiftLint))

rofle100lvl commented on docs/docs/guides/quick-start/add-dependencies.md:
```diff
@@ -120,4 +120,4 @@ struct ContentView_Previews: PreviewProvider {
 
-Run the app from Xcode, and you should see the image loaded from the URL.
\ No newline at end of file
+Run the app from Xcode, and you should see the image loaded from the URL.
```
idk why this change is here

rofle100lvl commented on docs/docs/guides/develop/verification/implicit-dependencies.md:
```diff
@@ -0,0 +1,8 @@
+
+# Implicit dependencies
+
+Run the command `tuist lint implicit-imports` to lint the project for implicit imports. This command will scan all source files and display any implicit imports present in your codebase. To make an import explicit, simply add the imported module to the target in your Tuist project.
```
I'll add more info in next pr

fortmarek commented on Sources/TuistKit/Commands/LintImplicitImportsCommand.swift:
```diff
@@ -0,0 +1,27 @@
+import TuistSupport
+
+struct LintImplicitImportsCommand: AsyncParsableCommand {
+    init() {}
```
```suggestion
```

fortmarek commented on Sources/TuistKit/Commands/LintImplicitImportsCommand.swift:
```diff
@@ -0,0 +1,27 @@
+import ArgumentParser
+import Path
+import TuistLoader
```
```suggestion
```

I believe these imports are no longer necessary?

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,50 @@
+import XcodeGraph
+
+final class GraphImplicitImportLintService {
+    let targetScanner: TargetImportsScanning
```
```suggestion
    private let targetScanner: TargetImportsScanning
```

nit

fortmarek commented on Sources/TuistKit/Services/GraphImplicitImportLintService.swift:
```diff
@@ -0,0 +1,50 @@
+import TuistLoader
+import XcodeGraph
+
+final class GraphImplicitImportLintService {
```
Should this conform to the `GraphLinting` protocol? Otherwise, we can probably remove the Config from the function signature [here](https://github.com/tuist/tuist/pull/6516/files#diff-432d00a701dd55696c8c99219c1ea800d4fe63626e0a6a1d36d93e55f753dbdaR15)

fortmarek commented on Sources/TuistKit/Services/LintImplicitImportsService.swift:
```diff
@@ -0,0 +1,57 @@
+}
+
+final class LintImplicitImportsService {
+    private let graphImplicitLintService: GraphImplicitImportLintService
```
The `GraphImportLintService` should be included here as a protocol, not the implementation

fortmarek commented on Sources/TuistKit/Services/TargetImportsScanner.swift:
```diff
@@ -0,0 +1,51 @@
+
+final class TargetImportsScanner: TargetImportsScanning {
+    private let importSourceCodeScanner: ImportSourceCodeScanner
+    private let fileSystem = FileSystem()
```
This should be included in the `init` as all the other dependencies

fortmarek commented on docs/docs/guides/develop/verification.md:
```diff
@@ -0,0 +1,9 @@
+---
+title: Verification
+description: Learn how to use Tuist to verificate your projects efficiently.
```
```suggestion
description: Learn how to use Tuist to verify your projects efficiently.
```

fortmarek reviewed (commented):
Just a couple of nits, after which we should be good to merge from my side

alexanderwe commented:
Just saw this PR and wanted to say how awesome it is that we can get this via linter within tuist 👍 Just a question to this comment here: https://github.com/tuist/tuist/pull/6516#issuecomment-2308677621 will this then be part of executing `tuist inspect implicit-dependencies` or when do these error be visible in the project ?

rofle100lvl commented:
> Just saw this PR and wanted to say how awesome it is that we can get this via linter within tuist 👍 Just a question to this comment here: [#6516 (comment)](https://github.com/tuist/tuist/pull/6516#issuecomment-2308677621) will this then be part of executing `tuist inspect implicit-dependencies` or when do these error be visible in the project ?
For sure, interface would be like that. Xcode -> Run Script phase -> tuist inspect implicit-dependencies --format xcode -> then warning will appear in the project

sphanley commented:
Hey @pepicrft: just a heads up that the 4.26 changelog lists the new command as being `tuist inspect implicit-dependencies` as per this PR title, but it seems like it's in fact `tuist inspect implicit-imports`. Not sure if it's appropriate to amend the changelog or anything but it may cause confusion.
